### PR TITLE
Simplify map conditional compilation

### PR DESCRIPTION
**Pull request by mulmarta**
No functional changes. The PR removes a bunch of conditional compilation flags related to deciding on the type of map that is used: HashMap, Vec or BTreeMap.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT license.

### CODE REVIEW

**Review by tomleavy — APPROVED** [review]

File: mls-rs/src/map.rs [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+        pub fn remove(&mut self, key: &K) -> Option<V> {
+            self.find(key).map(|i| self.0.remove(i).1)
+        }
+
+        fn find(&self, key: &K) -> Option<usize> {
+            self.0
+                .iter()
+                .find_position(|(k, _)| k == key)
+                .map(|(i, _)| i)
```
Comment (stefunctional) [inline mls-rs/src/map.rs:68]: [Nit]
```suggestion
                .position(|(k, _)| k == key)
```

File: mls-rs/src/map.rs [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+use mls_rs_codec::{MlsDecode, MlsEncode, MlsSize};
+
+#[cfg(feature = "std")]
+mod map_impl {
+    use core::hash::Hash;
+    use std::collections::{hash_map::Entry, HashMap};
+
+    #[derive(Clone, Debug, PartialEq, Eq)]
+    #[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+    pub struct SmallMap<K: Hash + Eq + PartialEq, V>(pub(super) HashMap<K, V>);
```
Comment (stefunctional) [inline mls-rs/src/map.rs:22]: [Nit] `Eq` implies `PartialEq`.

**Review by stefunctional — APPROVED** [review]
Just 2 small nits that are really more for your information; no need to change.

Thank you for cleaning up this map code!

File: mls-rs/src/map.rs (hunk 2 above)
Comment (mulmarta) [inline mls-rs/src/map.rs:22]:
```suggestion
    pub struct SmallMap<K: Hash + Eq, V>(pub(super) HashMap<K, V>);
```

File: mls-rs/src/map.rs [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+    use alloc::{
+        collections::{btree_map::Entry, BTreeMap},
+        vec::Vec,
+    };
+    #[cfg(feature = "by_ref_proposal")]
+    use itertools::Itertools;
+
+    #[derive(Clone, Debug, PartialEq, Eq)]
+    pub struct SmallMap<K: Hash + Eq + PartialEq, V>(pub(super) Vec<(K, V)>);
```
Comment (mulmarta) [inline mls-rs/src/map.rs:41]:
```suggestion
    pub struct SmallMap<K: Hash + Eq, V>(pub(super) Vec<(K, V)>);
```

File: mls-rs/src/map.rs [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+    pub type LargeMap<K, V> = BTreeMap<K, V>;
+    pub(super) type SmallMapInner<K, V> = Vec<(K, V)>;
+    pub type LargeMapEntry<'a, K, V> = Entry<'a, K, V>;
+
+    #[cfg(feature = "by_ref_proposal")]
+    impl<K: Hash + Eq + PartialEq, V> SmallMap<K, V> {
```
Comment (mulmarta) [inline mls-rs/src/map.rs:48]:
```suggestion
    impl<K: Hash + Eq, V> SmallMap<K, V> {
```

File: mls-rs/src/map.rs [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        fn find(&self, key: &K) -> Option<usize> {
+            self.0
+                .iter()
+                .find_position(|(k, _)| k == key)
+                .map(|(i, _)| i)
+        }
+    }
+}
+
+impl<K: Hash + Eq + PartialEq, V> Default for SmallMap<K, V> {
```
Comment (mulmarta) [inline mls-rs/src/map.rs:73]:
```suggestion
impl<K: Hash + Eq, V> Default for SmallMap<K, V> {
```

File: mls-rs/src/map.rs [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    fn default() -> Self {
+        Self(SmallMapInner::new())
+    }
+}
+
+impl<K: Hash + Eq + PartialEq, V> Deref for SmallMap<K, V> {
```
Comment (mulmarta) [inline mls-rs/src/map.rs:79]:
```suggestion
impl<K: Hash + Eq, V> Deref for SmallMap<K, V> {
```

File: mls-rs/src/map.rs [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    type Target = SmallMapInner<K, V>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<K: Hash + Eq + PartialEq, V> DerefMut for SmallMap<K, V> {
```
Comment (mulmarta) [inline mls-rs/src/map.rs:87]:
```suggestion
impl<K: Hash + Eq, V> DerefMut for SmallMap<K, V> {
```

File: mls-rs/src/map.rs [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl<K, V> MlsDecode for SmallMap<K, V>
+where
+    K: Hash + Eq + PartialEq + MlsEncode + MlsDecode + MlsSize,
```
Comment (mulmarta) [inline mls-rs/src/map.rs:95]:
```suggestion
    K: Hash + Eq + MlsEncode + MlsDecode + MlsSize,
```

File: mls-rs/src/map.rs [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+    V: MlsEncode + MlsDecode + MlsSize,
+{
+    fn mls_decode(reader: &mut &[u8]) -> Result<Self, mls_rs_codec::Error> {
+        SmallMapInner::mls_decode(reader).map(Self)
+    }
+}
+
+impl<K, V> MlsSize for SmallMap<K, V>
+where
+    K: Hash + Eq + PartialEq + MlsEncode + MlsDecode + MlsSize,
```
Comment (mulmarta) [inline mls-rs/src/map.rs:105]:
```suggestion
    K: Hash + Eq + MlsEncode + MlsDecode + MlsSize,
```

File: mls-rs/src/map.rs [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+    V: MlsEncode + MlsDecode + MlsSize,
+{
+    fn mls_encoded_len(&self) -> usize {
+        self.0.mls_encoded_len()
+    }
+}
+
+impl<K, V> MlsEncode for SmallMap<K, V>
+where
+    K: Hash + Eq + PartialEq + MlsEncode + MlsDecode + MlsSize,
```
Comment (mulmarta) [inline mls-rs/src/map.rs:115]:
```suggestion
    K: Hash + Eq + MlsEncode + MlsDecode + MlsSize,
```

**Review by stefunctional — DISMISSED** [review]

**Review by stefunctional — APPROVED** [review]

**Review by tomleavy — APPROVED** [review]

**Pull request merged**